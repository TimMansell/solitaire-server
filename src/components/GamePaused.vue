<template>
  <GameOverlay center-content show-logo alt v-if="showOverlay" data-test="game-paused">
    <template #title>Game Paused</template>
    <template #msg v-if="!isGamePaused.isActive">
      Your game has been paused due to inactivity
    </template>
    <template #buttons>
      <PauseGameButton />
    </template>
  </GameOverlay>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import GameOverlay from '@/components/GameOverlay.vue';
import PauseGameButton from './PauseGameButton.vue';

export default {
  name: 'GamePaused',
  components: {
    GameOverlay,
    PauseGameButton,
  },
  computed: {
    ...mapGetters(['isGamePaused']),
    showOverlay() {
      const { isPaused, showMsg } = this.isGamePaused;

      return isPaused && showMsg;
    },
  },
  mounted() {
    window.addEventListener(
      'visibilitychange',
      (e) => setTimeout(this.checkGamePaused, 2000, e),
      false
    );
  },
  destroyed() {
    window.removeEventListener('visibilitychange', this.checkGamePaused);
  },
  methods: {
    ...mapActions(['setGamePaused']),
    checkGamePaused(e) {
      const { isPaused } = this.isGamePaused;
      const { visibilityState } = e.target;

      if (isPaused) return;

      if (visibilityState === 'hidden') {
        const options = {
          showMsg: true,
        };

        this.setGamePaused(options);
      }
    },
  },
};
</script>
