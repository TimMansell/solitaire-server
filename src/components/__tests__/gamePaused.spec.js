import { shallowMount } from '@vue/test-utils';
import GamePaused from '@/components/GamePaused.vue';

describe('GamePaused.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(GamePaused, {
      computed: {
        isGamePaused: () => ({
          paused: true,
          active: true,
        }),
        hasMoves: () => true,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('shows inactive msg', () => {
    const wrapper = shallowMount(GamePaused, {
      computed: {
        isGamePaused: () => ({
          paused: true,
          active: false,
        }),
        hasMoves: () => true,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
