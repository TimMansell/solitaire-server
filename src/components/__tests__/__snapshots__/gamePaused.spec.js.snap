// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePaused.vue hides overlay 1`] = ``;

exports[`GamePaused.vue matches snapshot 1`] = `
<gameoverlay-stub alt="true" centercontent="true" showlogo="true" data-test="game-paused">Game Paused<pausegamebutton-stub></pausegamebutton-stub>
</gameoverlay-stub>
`;

exports[`GamePaused.vue shows inactive msg 1`] = `
<gameoverlay-stub alt="true" centercontent="true" showlogo="true" data-test="game-paused">Game Paused
  Your game has been paused due to inactivity
  <pausegamebutton-stub></pausegamebutton-stub>
</gameoverlay-stub>
`;
